### PR TITLE
Syntax highlighting for C

### PR DESCRIPTION
**Pull request by karupayun**
This PR creates the Syntax highlighter for source code (C++).
For now, the syntax highlighter is only using C-rules. It will be
expanded with C++ rule in a future PR.

It can be tested using CodeViewerDemo. There you will see a few fake unit-tests only to make regex more understandable and easier to modify in the future. 
fake unit tests

Also there is a copy of CaptureServiceImpl.cpp as an
example to test how the Syntax Highlighter works with a real file.
real example

Related to http://b/177651839.

### CODE REVIEW

File: src/CodeViewerDemo/CodeExamples.h
Comment (beckerhe) [inline src/CodeViewerDemo/CodeExamples.h:5]: QString?

File: src/CodeViewerDemo/CodeExamples.h
Comment (beckerhe) [inline src/CodeViewerDemo/CodeExamples.h:7]: If you define these variables in a header you have to declare them `inline`. Otherwise you will get linker error when including them from different cpp-files.

File: src/SyntaxHighlighter/Cpp.cc
Comment (beckerhe) [inline src/SyntaxHighlighter/Cpp.cc:1]: We usually use the file extension `.cpp`. Can you rename the file?

File: src/SyntaxHighlighter/Cpp.cc
Comment (beckerhe) [inline src/SyntaxHighlighter/Cpp.cc:7]: `#include <QRegularExpression>`

The same in the header file.

File: src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h
Comment (beckerhe) [inline src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h:16]: Very, very nit: cpp -> C++

File: src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h
Comment (beckerhe) [inline src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h:24]: Do you need the explicit values? If not I would remove the `= 0`, `= 1`, and `= 2`.

Comment (beckerhe) [reply src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h:24]: And it seems you don't need this enum for the API. So I would move it to the .cpp-file into a anonymous namespace.

File: src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h
Comment (beckerhe) [inline src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h:21]: One more thing: Dimitry recently pointed out, that we usually use `//` also for block comments. I did this wrong in the assembly syntax highlighter. But we should probably follow that guidance.

File: src/SyntaxHighlighter/Cpp.cc
Comment (beckerhe) [inline src/SyntaxHighlighter/Cpp.cc:14]: It's fine to put these constants into a namespace with a name. But you still need the anonymous namespace as well.

File: src/SyntaxHighlighter/CMakeLists.txt
Comment (beckerhe) [inline src/SyntaxHighlighter/CMakeLists.txt:8]: Could you add these files to the `target_sources` directives below instead?

File: src/CodeViewerDemo/main.cpp
Comment (beckerhe) [inline src/CodeViewerDemo/main.cpp:17]: Nit: I would remove this variable and directly call SetSourceCode and pass `cpp_example` in.

File: src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h
Comment (beckerhe) [reply src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h:24]: And could you make it an `enum class`. From the code I've seen, you already scope all usages properly.

File: src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h
Comment (beckerhe) [inline src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h:43]: These are public and they shouldn't. (That's also wrong in the Assembly highlighter. Apologies. :see_no_evil: )

File: src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h
Comment (beckerhe) [inline src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h:8]: This header is not needed here. (Once more the same problem is in the assembly highlighter. I really should review my own code. :see_no_evil:  My bad. )

**Review by beckerhe — COMMENTED** [review]
Thank you Pablo. Looks already really good. Much more readable than uncolored code.

I added some minor comments. The overall direction looks good. No complaints there. Just a few C++-y things.

File: src/SyntaxHighlighter/Cpp.cc
Comment (florian-kuebler) [inline src/SyntaxHighlighter/Cpp.cc:82]: I guess long term, we want to have central place for all colors. For now, I think it is fine as it is.

**Review by florian-kuebler — APPROVED** [review]
That's pretty cool. Thanks. I share some of the comments from Henning. So see my review as LGTM as soon as he approved!

File: src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h
Comment (karupayun) [reply src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h:21]: Do you want me to also change these things in X86Assembly?

Comment (beckerhe) [reply src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h:21]: No. This was not my intention at all. Of course I won't stop you if you tackle this. But I already made a mental note to address this soonish.

File: src/SyntaxHighlighter/Cpp.cc
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cc]: Thanks! Do you like it now?

File: src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h
Comment (karupayun) [reply src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h]: Done to the first 2. I didn't make an enum class just because I have to convert to an int for setCurrentBlockState. Is there a nice way to do that?

File: src/CodeViewerDemo/CodeExamples.h
Comment (karupayun) [reply src/CodeViewerDemo/CodeExamples.h:7]: Yes, thanks!

File: src/SyntaxHighlighter/Cpp.cc
Comment (beckerhe) [reply src/SyntaxHighlighter/Cpp.cc]: Yes. Looks good!

File: src/CodeViewerDemo/main.cpp
Comment (karupayun) [reply src/CodeViewerDemo/main.cpp:17]: I supposed that was better to have that variable because we are using it twice

File: src/SyntaxHighlighter/Cpp.cc
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cc]: Yes, completely agree. Is also one of the things in my roadmap that I want to tackle.

File: src/CodeViewerDemo/main.cpp
Comment (beckerhe) [reply src/CodeViewerDemo/main.cpp:17]: Ah sorry. Yes. Sry...

File: src/SyntaxHighlighter/Cpp.cpp
Comment (ronaldfw) [inline src/SyntaxHighlighter/Cpp.cpp:89]: Ok, I'm really bad with reading regular expressions, so I'm sorry if I'm getting this wrong. 

It looks to me that at least the "include file" and "string" and "macro" regex work together in interesting ways here and that the order matters. 

"Include file regex" will turn green a string like '#include<header.h>'. "String regex" will turn green a string like '"this"' including '"header.h"'. And finally macro regex turns '#pragma', but also '#include', into yellow orange.

So, 

#include <file.h>
#include "header.h"

turn out the same, but due to different rules. 

Doing the "include file regex" after "macro regex" would also lead to a different result (example: #include with <> and "" would look different), right? That may be fine and intended, but I would add a comment to explain that.

File: src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h
Comment (beckerhe) [reply src/SyntaxHighlighter/include/SyntaxHighlighter/Cpp.h]: No, not really. You have to call `static_cast` all the time.

**Review by beckerhe — APPROVED** [review]
LGTM! Thanks Pablo!

File: src/SyntaxHighlighter/Cpp.cpp
Comment (ronaldfw) [inline src/SyntaxHighlighter/Cpp.cpp:43]: Should this better be called kPreprocessorRegex as you are detecting preprocessor directives and not macros (which may be indistinguishable from functions in general)?

File: src/SyntaxHighlighter/Cpp.cpp
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cpp:89]: There are two different things
1- Yes, I tried to follow the same rules than X86Assembly where we process line by line and apply the rules one on top of another. So, the order matters. If we have **"Hi world, I'm an int variable"**, first **"int"** is painted and after this is overpainted by the comment thing". This leads to one problem, what comes first? Comment or String?
_" /* Hi, I'm an string */. ""_
_/*  "Hi, I'm a comment" */_
Is impossible to solve with the approach, but as they are really rare, we are continuing with our approach. If not, a more complex process is needed. But if you feel that it is important, I can change it.
2- The fact with macro and string was intended. But now, thinking about it better, I'm realizing that I can avoid that without much effort. Will change it.

Comment (ronaldfw) [reply src/SyntaxHighlighter/Cpp.cpp:89]: As far as I can tell, the approach is fine for our purposes. I would suggest to document it a bit, at least a comment in the code that the order matters and shouldn't be changed unless one knows what they are doing.

File: src/SyntaxHighlighter/Cpp.cpp
Comment (antonrohr) [inline src/SyntaxHighlighter/Cpp.cpp:14]: There might be a reason that I don't see, but why this extra namespace here?

**Review by antonrohr — COMMENTED** [review]
Only one minor comment in the code and here comes another big-ish one. Feel free to not do it, because it is a bit of work and other people already gave their LGTM, but: How about unit tests? I feel one test for every regex might bring a lot of benefit. It protects us against future accidental changes and it documents what each regex does for the rest of us.

File: src/SyntaxHighlighter/Cpp.cpp
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cpp:14]: I wanted to have the C-namespace because after I want to include the C++ rules and Henning asked for the anonymous also. @beckerhe ??

File: src/SyntaxHighlighter/Cpp.cpp
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cpp:108]: Added a few comments in the regex and also when we're applying them. My goal is include at the first the "heuristic ones", where we can modify them following C++/Google conventions and to make the code feels a little bit nicer.

**Comment by karupayun** [comment]
> Only one minor comment in the code and here comes another big-ish one. Feel free to not do it, because it is a bit of work and other people already gave their LGTM, but: How about unit tests? I feel one test for every regex might bring a lot of benefit. It protects us against future accidental changes and it documents what each regex does for the rest of us.

I liked that suggestion, will think about that :)

File: src/SyntaxHighlighter/Cpp.cpp
Comment (beckerhe) [reply src/SyntaxHighlighter/Cpp.cpp:14]: `namespace C` for the name. And the unnamed `namespace` to get internal linkage.

Comment (beckerhe) [reply src/SyntaxHighlighter/Cpp.cpp:14]: There is no way to shorten that. At least none that I know of.

Comment (antonrohr) [reply src/SyntaxHighlighter/Cpp.cpp:14]: Got it, thank you

**Comment by karupayun** [comment]
So, there are a few more things here:
1- Update PR descriptions with a few screenshots
2- CodeViewerDemo shows "fake unit tests" for interesting regex (and also a real file after that)
3- Comments in every regex

**Pull request merged**
